server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MirroringTask.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.centraldogma.server.internal.mirror;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+
+import java.io.File;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.centraldogma.server.command.CommandExecutor;
+import com.linecorp.centraldogma.server.mirror.Mirror;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+
+final class MirroringTask {
+
+    private static Iterable<Tag> generateTags(Mirror mirror) {
+        return ImmutableList.of(
+                Tag.of("direction", mirror.direction().name()),
+                Tag.of("remoteRepo", mirror.remoteRepoUri().toString()),
+                Tag.of("remoteBranch", firstNonNull(mirror.remoteBranch(), "")),
+                Tag.of("remotePath", mirror.remotePath()),
+                Tag.of("localRepo", mirror.localRepo().name()),
+                Tag.of("localPath", mirror.localPath()));
+    }
+
+    private final MeterRegistry meterRegistry;
+    private final Mirror mirror;
+    private final Iterable<Tag> tags;
+
+    MirroringTask(Mirror mirror, MeterRegistry meterRegistry) {
+        this.mirror = mirror;
+        this.meterRegistry = meterRegistry;
+        tags = generateTags(mirror);
+    }
+
+    private Counter counter(boolean success) {
+        return Counter.builder("mirroring.result")
+                      .tags(tags)
+                      .tag("success", Boolean.toString(success))
+                      .register(meterRegistry);
+    }
+
+    void run(File workDir, CommandExecutor executor, int maxNumFiles, long maxNumBytes) {
+        try {
+            meterRegistry.timer("mirroring.task", tags)
+                         .record(() -> mirror.mirror(workDir, executor, maxNumFiles, maxNumBytes));
```

Comment:
I'm worried this can create too many metrics because the cardinality of the tags can grow easily. How about manually recording only 1) failed mirrors or 2) slow mirrors which has executed longer than a thread hold? I think 1) is mandatory, but 2) could be optional and ignored at the moment.

Reply:
If we only record certain events conditionally, I'm afraid we won't be able to distinguish between cases where:
1) The mirroring didn't run at all
2) The mirroring succeeded, but just isn't showing

I didn't think metric cardinality would be an issue mainly because
1) The number of mirroring tasks is likely to be limited
2) By design, CD relies on zookeeper quorum which limits the number of servers that can produce metrics.

Having said this, if you feel the number of metrics is still an issue, I guess a simpler solution would only to add metrics for the `mirroring-scheduler` and remove `MirroringTask` altogether. What do you think?

Reply:
I see. Changing the metrics would be not a difficult task so let's keep it as it is.